libs/nx-terraform/src/generators/create-environment-sp/generator.ts

```diff
@@ -0,0 +1,97 @@
+import { Tree } from '@nrwl/devkit'
+import execa from 'execa'
+import { isDryRun } from '../../common/isDryRun'
+import { readRepoSettings } from '../../common/read-repo-settings'
+import { readConfigFromEnvFile } from '../../common/readConfigFromEnvFile'
+import { NxTerraformCreateEnvironmentServicePrincipalSchema } from './schema'
+
+export default async function (
+    tree: Tree,
+    options: NxTerraformCreateEnvironmentServicePrincipalSchema,
+) {
+    const repoSettings = readRepoSettings()
+    const environmentConfig = await readConfigFromEnvFile(
+        repoSettings.terraformStateType,
+        options.environmentName,
+    )
+
+    const servicePrincipalName = `gh-actions-${repoSettings.azureResourcePrefix}-${options.environmentName}-sp`
+    const scopes = `/subscriptions/${environmentConfig.subscriptionId}/resourcegroups/${environmentConfig.resourceGroupName}`
+    const containerScope = `/subscriptions/${environmentConfig.subscriptionId}/resourceGroups/${environmentConfig.resourceGroupName}/providers/Microsoft.Storage/storageAccounts/${environmentConfig.terraformStorageAccount}/blobServices/default/containers/${environmentConfig.terraformStorageContainer}`
+
+    if (isDryRun()) {
+        console.log('Will run:')
+
+        console.log(`> az ad sp create-for-rbac --role ${options.role} \
```

Comment:
Maybe look into whether you need to specify **skip-assignment** flag? I feel that in the past if you didn't specify that then the SP gets subscription perms - but maybe because you're specifying scopes this won't happen.

Reply:
Yeah, my understanding is that by specifying the scopes you don't get the default scope of the subscription.